src/HHbbVV/processors/bbVVSkimmer.py

```diff
@@ -684,6 +692,373 @@ def process(self, events: ak.Array):
+                            vbfJets_mask * mass_jj_cut_sorted_pt * eta_jj_cut_sorted_pt
+                        )
+                        num_sorted_pt = ak.fill_none(ak.sum(vbfJets_mask_sorted_pt, axis=1), 0)
+
+                        # here we can print some information about the jets so that we can study the selections a bit.
+                        # jet_pairs = ak.cartesian({"reco": vbfJets, "gen": vbfGenJets[:,0:2]}) # this is only for unfiltered events
+                        jet_pairs = ak.cartesian(
+                            {"reco": vbfJets_sorted_pt[:, 0:2], "gen": vbfGenJets[:, 0:2]}
+                        )
+
+                        # Calculate delta eta and delta phi for each pair
+                        delta_eta = jet_pairs["reco"].eta - jet_pairs["gen"].eta
+                        delta_phi = np.pi - np.abs(
+                            np.abs(jet_pairs["reco"].phi - jet_pairs["gen"].phi) - np.pi
+                        )
+
+                        # Calculate delta R for each pair
+                        delta_R = np.sqrt(delta_eta**2 + delta_phi**2)
+
+                        # Apply a mask for a low delta R value
+                        mask_low_delta_R = delta_R < 0.4
+                        num_per_event = ak.sum(
+                            mask_low_delta_R, axis=-1
+                        )  # miscounts 0's since some are empty
+
+                        # Combine masks with logical 'and' operation
+                        total_mask = num_sorted_pt > 1
+
+                        # set event that fail to have 0 for num of events.
+                        num_per_event = ak.where(total_mask, num_per_event, 0)
+                        vbfVars[f"vbfR1{R1:.1f}R2{R2:.1f}"] = num_per_event.to_numpy()
+
+                        # note later when we have the pd dataframe, we need to apply mask, then turn data into graph.
+                        # and df = df[selection_mask & (df[('nGoodMuons', 0)] == 0) & (df[('nGoodElectrons', 0)] == 0) & (df[('nGoodVBFJets', 0)] >= 1)& (df[('nGoodJets', 0)] == 0)]
+
+        vbfVars[f"vbfpt"] = pad_val(vbfJets_sorted_pt.pt, 2, axis=1)
+        vbfVars[f"vbfeta"] = pad_val(vbfJets_sorted_pt.eta, 2, axis=1)
+        vbfVars[f"vbfphi"] = pad_val(vbfJets_sorted_pt.phi, 2, axis=1)
+        vbfVars[f"vbfM"] = pad_val(vbfJets_sorted_pt.mass, 2, axis=1)
+
+        # int list representing the number of passing vbf jets per event
+        vbfVars[f"nGoodVBFJets"] = n_good_vbf_jets_sorted_pt.to_numpy()
+
+        adding_bdt_vars = True
+        if adding_bdt_vars == True:
+            import time
+
+            # start_time = time.time()
+            # Adapted from HIG-20-005 ggF_Killer 6.2.2
+            # https://coffeateam.github.io/coffea/api/coffea.nanoevents.methods.vector.PtEtaPhiMLorentzVector.html
+            # https://coffeateam.github.io/coffea/api/coffea.nanoevents.methods.vector.LorentzVector.html
+
+            def to_four_momentum(pt, phi, eta, m):
```

Comment:
These functions are not necessary because `coffea` already has these "Mixin"s built-in which allow you to e.g. `vbf_quark1 + vbf_quark2` or `vbf_quark1.px` etc. directly. Although sometimes you do have to tell coffea to add these manually (I can help with that). Let me know if you're having trouble using the coffea built-in stuff for these on Slack.

Reply:
Ah I see, I wasn't sure if I could access anything other than mass eta... etc, thanks. Also, I've noticed when comparing the dijet invariant mass calculations using coffea methods vs vector method that the vector methods are more precise (more decimals displayed in output) so maybe the coffea methods use some rounding?

Reply:
Interesting, maybe they use different floating point precision by default. In general the `coffea` developers recommend using the `vector` library since it'll be better maintained, but only if it's convenient. For the analysis I don't think it matters.